frex/utils/constraint_solver_utils.py

```diff
@@ -0,0 +1,123 @@
+from ortools.sat.python import cp_model
+from ortools.linear_solver import pywraplp
+from frex.models import Candidate
+from typing import Tuple, Dict
+from dataclasses import fields
+
+
+class ConstraintSolverUtils:
+    """
+    A utility class for general constraint solving related to recommendations.
+    """
+
+    @staticmethod
+    def solve_constraints_on_candidates(
+        *,
+        candidates: Tuple[Candidate],
```

Comment:
Rather than have this C-like procedural interface I'd rather see a fluent interface / builder pattern like

ConstraintSolver().set_candidates(...).set_sections(...).add_section_constraint(...).solve()

Then each of the setters can have documentation as to what exactly that parameter means (i.e., section).